builder/dockerfile/evaluator.go

```diff
@@ -103,13 +103,14 @@ func dispatch(d dispatchRequest, cmd instructions.Command) (err error) {
 
 // dispatchState is a data object which is modified by dispatchers
 type dispatchState struct {
-	runConfig  *container.Config
-	maintainer string
-	cmdSet     bool
-	imageID    string
-	baseImage  builder.Image
-	stageName  string
-	buildArgs  *buildArgs
+	runConfig       *container.Config
+	maintainer      string
+	cmdSet          bool
+	imageID         string
+	baseImage       builder.Image
+	stageName       string
+	buildArgs       *buildArgs
+	operatingSystem string
```

Comment:
Is this really needed? `exportImage` for example looks up the same value by `imageID` (or I think `baseImage` could be also used here). If a new value like this is added it should be typed to specific supported values, as this value is not abstract but only certain values are supported by the code.

Reply:
It is required. I can go through again, maybe I missed one, but the value is validated when it's parsed in the dockerfile (currently the parser directive, but updated to `FROM --platform=...` in https://github.com/moby/moby/pull/35089. Let me see if I can remove it.

Reply:
The entrypoints into the builder where the operating system is passed is through BuildFromConfig in the `daemon.Import` and `daemon.Import` paths. In both these cases, this already should have been validated. But for additional safety, I have added a check using the existing `system.IsOSSupported` call, even though I believe it's overly paranoid.

Similarly I have added a similar check in `newDispatchState` to ensure it is always set to a valid value prior to calling a dispatch steps. The previous check in `beginStage` has also been left in place to cater for multi-stage (not that we can currently cope with switching operating systems in multi-stage, or at least copying a file from an LCOW container into a WCOW container specifically).

The remaining items in the daemon which are indexed by operating system at this point are the graphDrivers and the layerStores in the main daemon structure, not the other stores. The builder doesn't interact directly with the graphDriver, but does the layerstore through Commit. With these additional safety checks, I believe this is now completely safe. With the use of this validation, I don't believe it needs an explicit type, the same way that things such as the ImageID don't have an explicit type in the builder structures (although I realise that isn't used as an index in quite the same way).

Have just pushed this update so the builder should be completely safe now from panic through invalid index.